plugin-server/src/main/ingestion-queues/session-recording-v2/retention/retention-aware-batch-writer.ts

```diff
@@ -0,0 +1,93 @@
+import { S3Client } from '@aws-sdk/client-s3'
+
+import { ValidRetentionPeriods } from '../constants'
+import { S3SessionBatchFileStorage } from '../sessions/s3-session-batch-writer'
+import {
+    SessionBatchFileStorage,
+    SessionBatchFileWriter,
+    SessionData,
+    WriteSessionResult,
+} from '../sessions/session-batch-file-storage'
+import { RetentionPeriod } from '../types'
+import { RetentionService } from './retention-service'
+
+class RetentionAwareBatchFileWriter implements SessionBatchFileWriter {
+    private storageMap: { [key in RetentionPeriod]: SessionBatchFileStorage }
+    private writerMap: { [key in RetentionPeriod]: SessionBatchFileWriter | null }
+
+    constructor(
+        private readonly s3: S3Client,
+        private readonly bucket: string,
+        private readonly prefix: string,
+        private readonly timeout: number,
+        private readonly retentionService: RetentionService
+    ) {
+        this.storageMap = ValidRetentionPeriods.reduce(
+            (storage, retentionPeriod) => {
+                storage[retentionPeriod] = new S3SessionBatchFileStorage(
+                    this.s3,
+                    this.bucket,
+                    retentionPeriod === 'legacy' ? 'session_recording_batches' : `${this.prefix}/${retentionPeriod}`,
+                    this.timeout
+                )
+                return storage
+            },
+            {} as { [key in RetentionPeriod]: SessionBatchFileStorage }
+        )
```

Comment:
nit: The storage map could be created once in `RetentionAwareStorage` instead of creating it on every batch. We'd also be able to implement the healthcheck there.